x-pack/plugin/ccr/qa/chain/src/test/java/org/elasticsearch/xpack/ccr/ChainIT.java

```diff
@@ -0,0 +1,176 @@
+                        "\"_doc\": {" +
+                                "  \"_source\": {" +
+                                "    \"includes\": [\"field\"]," +
+                                "    \"excludes\": [\"filtered_field\"]" +
+                                "   }"+
+                                "}";
+            }
+            Settings indexSettings = Settings.builder()
+                    .put("index.soft_deletes.enabled", true)
+                    .build();
+            createIndex(leaderIndexName, indexSettings, mapping);
+            for (int i = 0; i < numDocs; i++) {
+                logger.info("Indexing doc [{}]", i);
+                index(client(), leaderIndexName, Integer.toString(i), "field", i, "filtered_field", "true");
+            }
+            refresh(leaderIndexName);
+            verifyDocuments(leaderIndexName, numDocs);
+        } else if ("middle".equals(targetCluster)) {
+            logger.info("Running against middle cluster");
+            followIndex("leader_cluster:" + leaderIndexName, middleIndexName);
+            assertBusy(() -> verifyDocuments(middleIndexName, numDocs));
+            try (RestClient leaderClient = buildLeaderClient()) {
+                int id = numDocs;
+                index(leaderClient, leaderIndexName, Integer.toString(id), "field", id, "filtered_field", "true");
+                index(leaderClient, leaderIndexName, Integer.toString(id + 1), "field", id + 1, "filtered_field", "true");
+                index(leaderClient, leaderIndexName, Integer.toString(id + 2), "field", id + 2, "filtered_field", "true");
+            }
+            assertBusy(() -> verifyDocuments(middleIndexName, numDocs + 3));
+        } else if ("follow".equals(targetCluster)) {
+            logger.info("Running against follow cluster");
+            final String followIndexName = "follow";
+            followIndex("middle_cluster:" + middleIndexName, followIndexName);
+            assertBusy(() -> verifyDocuments(followIndexName, numDocs + 3));
+
+            try (RestClient leaderClient = buildLeaderClient()) {
+                int id = numDocs + 3;
+                index(leaderClient, leaderIndexName, Integer.toString(id), "field", id, "filtered_field", "true");
+                index(leaderClient, leaderIndexName, Integer.toString(id + 1), "field", id + 1, "filtered_field", "true");
+                index(leaderClient, leaderIndexName, Integer.toString(id + 2), "field", id + 2, "filtered_field", "true");
+            }
+
+            try (RestClient middleClient = buildMiddleClient()) {
+                assertBusy(() -> verifyDocuments(middleIndexName, numDocs + 6, middleClient));
+            }
+
+            assertBusy(() -> verifyDocuments(followIndexName, numDocs + 6));
+        } else {
+            fail("unexpected target cluster [" + targetCluster + "]");
+        }
+    }
+
+    private static void index(RestClient client, String index, String id, Object... fields) throws IOException {
```

Comment:
Should we in a follow up add a Gradle module inside the qa directory that contains these helper methods and let the qa modules depend on this module? We've been repeating these methods many times now.

Reply:
Yes, I thought the same thing! Can you take that on?

Reply:
👍  I will work on it.